### DISCUSSION TITLE
Update screwed up Reach! Help!

### DISCUSSION THREAD

**Discussion by JamesLampert** [reach-sh/reach-lang#1444] [Q&A] [answered]
Two weeks ago today, discovering that the new "RSVP" tutorial requires features not in my then-current version of Reach, I did a
`./reach update`

At the time the version I had was working just fine. After I updated, it stopped working completely. If I do a
`./reach version`
(or a ./reach _anything else_, for that matter), something flickers in my Docker window, too fast to read, and then I get:
```
Jamess-iMac:tut jameslampert$ ./reach version
Missing:

reach 0.1.11 (98ccf409) - Reach command-line tool

Usage: reach COMMAND

Available options:
  -h,--help                Show this help text

Available commands:
  clean                    Delete 'build/$MODULE.$IDENT.mjs'
  compile                  Compile an app
  config                   Configure default Reach settings
  devnet                   Run only the devnet
  docker-reset             Kill and remove all Docker containers
  down                     Halt all Dockerized Reach services and devnets
  hashes                   Display git hashes used to build each Docker image
  help                     Show usage
  info                     List available updates
  init                     Set up source files for a simple app in the current
                           directory
  react                    Run a simple React app
  rpc-run                  Run an RPC server + frontend with development
                           configuration
  rpc-server               Run a simple Reach RPC server
  run                      Run a simple app
  scaffold                 Set up Docker scaffolding for a simple app in the
                           current directory
  support                  Create a GitHub gist of index.rsh and index.mjs
                           (default), or specify your own files to upload!
  tsa                      Run the TEAL Static Analyzer (only available with
                           Reach Everest)
  update                   Perform available updates
  version                  Display version

```

I'm on a Mac, running Catalina, VSCode is my Reach development environment, and Bash is my default command line shell. I _think_ I did the ./reach update from a terminal session within VSCode, but I get the subsequent utterly uninformative error message, and the helptext, regardless of whether I'm in a VSCode terminal session or a terminal window opened from Finder.

I've had this question up on Discord since the day it happened.

**Comment by jeapostrophe** [comment, marked as answer]
I am pretty sure the problem is needing a new cli image. If you delete the old one, it will download a new one, or you could specifically get a new one

@DanBurton can you copy in the Docker command to get a new cli image?

**Reply by JamesLampert** [reply]
Could be. Out of morbid curiosity, what's a "cli image"?

Hmm. I just clicked on "Images" in the left column of the Docker window. Hmm. There are 159 images, most of them with a name of "<none>," three with a name of "postgres," six "reachsh/devnet-algo," six "reachsh/devnet-cfx," six "reachsh/devnet-eth," six "reachsh/reach," seven with "reachsh/reach-app-x," where x is the name of a tutorial or workshop project, six "reachsh/reach-cli," 
six "reachsh/react-runner," six "reachsh/rpc-server," six "reachsh/runner," one "sandbox_algod" (marked "in use") and one "sandbox_indexer" (also marked "in use").

At any rate, I'm eagerly awaiting what to do next.

**Reply by jeapostrophe** [reply]
Nothing bad will happen if you delete all of those. You'll just re download the ones you need

The cli image is a Docker container that has the 'reach' command in it. It implements things like 'run' while the compiler itself is in the 'reachc' image

**Reply by JamesLampert** [reply]
> . . . while the compiler itself is in the 'reachc' image

Interesting. And "reachc" images are conspicuously absent.

**Reply by jeapostrophe** [reply]
Oh, they might be called 'reach'

**Reply by JamesLampert** [reply]
I deleted all the "reach-cli" images, and trying `./reach version` again. I got:

```
Jamess-iMac:tut jameslampert$ ./reach version
latest: Pulling from reachsh/reach-cli
59bf1c3509f3: Already exists 
94bb4c2c3f30: Pull complete 
7723b6e3663d: Pull complete 
d9c88eb0cc50: Pull complete 
b2d7d81f9db2: Pull complete 
a12eeae80ac0: Pull complete 
Digest: sha256:b8b44c6c16baf4bf4b0998cbf5535165bfdd752f4953c759c93c928205d8bfbc
Status: Downloaded newer image for reachsh/reach-cli:latest
docker.io/reachsh/reach-cli:latest
Missing:

reach 0.1.12 (816420cc) - Reach command-line tool

Usage: reach COMMAND

Available options:
  -h,--help                Show this help text

Available commands:
  clean                    Delete 'build/$MODULE.$IDENT.mjs'
  compile                  Compile an app
. . .
```
. . . _and so forth, as before_ . . .

I'll take this up again tomorrow: tonight's the Los Angeles Philharmonic's "John Williams Gala" concert (even the _cheap_ seats are as expensive as box seats at Hollywood Bowl, and I'm _in_ one of the cheap seats), so I'm leaving the office 2 hours early.

**Reply by DanBurton** [reply]
> @DanBurton can you copy in the Docker command to get a new cli image?

Sure. @JamesLampert can you try the following and share the output if something goes wrong?

```
docker pull reachsh/reach-cli:latest
./reach update
./reach hashes
```

**Reply by JamesLampert** [reply]
By the time I saw that last night, I'd already shut down Docker, and was in the process of shutting down my Mac. The concert was worth the ticket price, and as I was crossing Grand Avenue to get to Disney Hall, Maestro Dudamel was also crossing, and as we passed, we exchanged pleasantries. Our entire conversation consisted of five words: "Good evening, Maestro"; "Good evening." Not something that happens every concert night.

But back to the matter at hand, I'll let you know what happens as soon as Docker comes up.

**Reply by JamesLampert** [reply]
Tried it twice. Both times the same:
```
Jamess-iMac:tut jameslampert$ docker pull reachsh/reach-cli:latest
latest: Pulling from reachsh/reach-cli
59bf1c3509f3: Already exists 
94bb4c2c3f30: Pull complete 
7723b6e3663d: Pull complete 
d9c88eb0cc50: Pull complete 
b2d7d81f9db2: Pull complete 
a12eeae80ac0: Pull complete 
Digest: sha256:b8b44c6c16baf4bf4b0998cbf5535165bfdd752f4953c759c93c928205d8bfbc
Status: Downloaded newer image for reachsh/reach-cli:latest
docker.io/reachsh/reach-cli:latest
Jamess-iMac:tut jameslampert$ ./reach update
Missing:

reach 0.1.12 (816420cc) - Reach command-line tool

Usage: reach COMMAND

Available options:
  -h,--help                Show this help text

Available commands:
  clean                    Delete 'build/$MODULE.$IDENT.mjs'
  compile                  Compile an app
  config                   Configure default Reach settings
. . .
```
It sat there for several seconds on the `./reach update` before coming back with the usual cryptic error message and helptext.

**Reply by jeapostrophe** [reply]
Can you delete the `reach` script and re-download it?

**Reply by JamesLampert** [reply]
I'm assuming we're talking about the "reach" shell script sitting in each project library.

I have no doubt I could delete it and re-download it if I knew where to find it, and what tool to use for the download.

Please understand that I'm as much of a newbie here as you would likely be if somebody plopped you down in front of a terminal connected to an AS/400, with nothing on it but PDM, SEU, and an ILE RPG compiler, and told you to start coding. That's why I found Ms. Hodge's video tutorial series so valuable, and why most of the Workshops have me scratching my head.

**Reply by jeapostrophe** [reply]
```
rm reach
curl https://docs.reach.sh/reach -o reach
chmod +x reach
```

from https://docs.reach.sh/quickstart/

**Reply by JamesLampert** [reply]
Thanks. The "curl" call came back with
`curl: (60) SSL certificate problem: certificate has expired`

Is there a "do it anyway" switch for curl? Or can I do it from a browser, where I know how to override a cert problem?

**Reply by jeapostrophe** [reply]
I just checked that command on another machine and I think that something may be off with your machine/Internet which is giving you the bad certificate. But, the option is `-k`, so `curl -k https://docs.reach.sh/reach -o reach`

**Reply by JamesLampert** [reply]
That worked. `./reach version` returns `reach 0.1.12`

Running `./reach update` now.

**Reply by jeapostrophe** [reply]
Phew! I think you're probably good to go

**Reply by JamesLampert** [reply]
forgot how long it takes.  Now, I get:
```
Jamess-iMac:tut jameslampert$ ./reach version
reach 0.1.12
Jamess-iMac:tut jameslampert$ ./reach hashes
reach-cli: 816420cc
reach: 816420cc
react-runner: 816420cc
rpc-server: 816420cc
runner: 816420cc
devnet-algo: 816420cc
devnet-cfx: 816420cc
devnet-eth: 816420cc
```

I'm guessing that at this point, I just need to propagate (or maybe alias?) the new "reach" shell script to all other occurrences, and I can get right back into the RSVP tutorial.

**Reply by jeapostrophe** [reply]
Yup

**Reply by JamesLampert** [reply]
Interesting: I now have VSCode up, with the first (very incomplete) version of the "rsvp" project up.  When it came up, my Docker window showed a single container/app with a red icon. When I did a ./reach compile from VSCode's terminal pane, I got
```
Jamess-iMac:rsvp-1 jameslampert$ ./reach compile
sed: illegal option -- r
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
Verifying knowledge assertions
Verifying for generic connector
  Verifying when ALL participants are honest
  Verifying when NO participants are honest
Checked 9 theorems; No failures!
```
and the Docker window, the next time I looked at Containers/Apps, showed a whole bunch more than I can recall seeing before:
![dockerwindow28sp2022](https://user-images.githubusercontent.com/3721417/192868886-7154734e-1eee-452e-838c-4c0c780019e2.png)
Is something still wrong? What's the "sed:illegal option" about? I don't get it in a regular terminal window; it goes straight to "Verifying knowledge assertions."

**Reply by jeapostrophe** [reply]
I looked through all of the uses of `sed` in the code you're calling and I don't see any with `-r` in them, so I am confused. It apparently didn't cause a problem.

We expect to use one more container than before, but I don't expect to see so many. I'll have to investigate. Can you say what version of Docker you're running?

**Reply by JamesLampert** [reply]
Re: "sed" (for some reason I'm thinking of the "sed-a-give" gag from Young Frankenstein), the error message only happens in the VSCode terminal pane, so that might have something to do with it.

Docker Desktop 4.4.2 (73305)
Engine: 20.10.12
Compose: v2.2.3
Credential Helper: 0.6.4
Kubernetes: v1.22.5
Snyk: v1.809.0

VSCode is 
Version: 1.71.2 (Universal)
Commit: 74b1f979648cc44d385a2286793c226e611f59e7
Date: 2022-09-14T21:05:37.721Z (2 wks ago)

Not sure where to look, to find out what version of Reach support I have in VSCode, assuming that's relevant.
